### PR TITLE
Add AZ Support for VPN & ER GW PIPs

### PR DESCRIPTION
**Pull request by jhajduk-microsoft**
# Overview/Summary
This PR solves issue 241 for Zone-Redundant PIPs for Hub Networking Resources

## This PR fixes/adds/changes/removes

1. Adds ZR option for Firewall and Firewall PIP
2. Adds ZR option for Gateway PIPs
3. Adds new parameter for ZR gateway PIPs: parAzGatewayAvailabilityZones
4. Adds ability to deploy gateway options in any available zonal configuration separately

### Breaking Changes

1. New Parameter: parAzErGatewayAvailabilityZones
2. New Parameter: parAzVpnGatewayAvailabilityZones

## Testing Evidence
Will share permissions

## As part of this Pull Request I have

- [ X] Read the [Contribution Guide](https://github.com/Azure/ALZ-Bicep/wiki/Contributing) and ensured this PR is compliant with the guide
- [X ] Ensured the resource API versions in `.bicep` file/s I am adding/editing are using the latest API version possible
- [ X] Checked for duplicate [Pull Requests](https://github.com/Azure/ALZ-Bicep/pulls)
- [ ] Associated it with relevant [GitHub Issues](https://github.com/Azure/ALZ-Bicep/issues)
- [ ] *(ALZ Bicep Core Team Only)* Associated it with relevant [ADO Items](https://aka.ms/alz/bicep/backlog)
- [ X] Ensured my code/branch is up-to-date with the latest changes in the `main` [branch](https://github.com/Azure/ALZ-Bicep/tree/main)
- [X] Performed testing and provided evidence.
- [ ] Updated one or more of the following tests *(if required)*
  - [Unit](https://github.com/Azure/ALZ-Bicep/blob/main/.github/workflows/bicep-build-to-validate.yml)
  - [Linting](https://github.com/Azure/ALZ-Bicep/tree/main/.github/workflows)
  - [E2E (End-To-End)](https://github.com/Azure/ALZ-Bicep/blob/main/tests/pipelines/bicep-build-to-validate.yml)
  - [ValidateAzCloud (Base validation in Azure Cloud)](https://github.com/Azure/ALZ-Bicep/blob/main/tests/pipelines/base-unit-validate.yml)
  - [ValidateMcCloud (Base validation in Azure China Cloud)](https://github.com/Azure/ALZ-Bicep/blob/main/tests/pipelines/mc-base-unit-validate.yml)
- [ X] Updated relevant and associated documentation (e.g. Contribution Guide, Module READMEs, Wiki Docs etc.)
- [ ] If relevant, created or updated Code Tours [here](https://github.com/Azure/ALZ-Bicep/blob/main/.vscode/tours)

### CODE REVIEW

File: infra-as-code/bicep/modules/hubNetworking/hubNetworking.bicep [hunk 1]
```diff
@@ -74,9 +74,19 @@ param parAzFirewallTier string = 'Standard'
...
 ])
 @description('Availability Zones to deploy the Azure Firewall across. Region must support Availability Zones to use. If it does not then leave empty.')
-param parAzFirewallAvailabilityZones array = []
+param parAzFirewallAvailabilityZones array 
+
+@allowed([
+  '1'
+  '2'
+  '3'
+  '1,2,3'
```
Comment (olljanat) [inline infra-as-code/bicep/modules/hubNetworking/hubNetworking.bicep:86]: I think that this is not needed as parameter type is array?

**Review by olljanat — COMMENTED** [review]
@jhajduk-microsoft I just noticed this PR here which is touching same area as my draft PR #302 which other why fine but we probably need agree which is common approach for these.

cc @jtracey93 you probably want comment both of these?

File: infra-as-code/bicep/modules/hubNetworking/hubNetworking.bicep (hunk 1 above)
Comment (jhajduk-microsoft) [reply infra-as-code/bicep/modules/hubNetworking/hubNetworking.bicep:86]: @olljanat - I was unable either to deploy to multiple zones without adding it in explicitly. If you are able to deploy this module to multiple zones without explicitly defining it, then I will defer. If your PR covers greater breadth, then I can just remove this one in favor of yours. I will also be on the Bicep call on Thursday and we can discuss, if needed.

Comment (jhajduk-microsoft) [reply infra-as-code/bicep/modules/hubNetworking/hubNetworking.bicep]: You are correct, thanks for catching this. I removed the explicit 1,2,3 from the allowed values
@jtracey93 - This should be ready now

**Comment by jhajduk-microsoft** [comment]
@olljanat - Fixed, please review

File: infra-as-code/bicep/modules/privateDnsZones/privateDnsZones.bicep [hunk 2]
```diff
@@ -136,7 +136,9 @@ var varAzBackupGeoCodes = {
 }
 
 // If region entered in parLocation and matches a lookup to varAzBackupGeoCodes then insert Azure Backup Private DNS Zone with appropriate geo code inserted alongside zones in parPrivateDnsZones. If not just return parPrivateDnsZones
-var varPrivateDnsZonesMerge = contains(varAzBackupGeoCodes, parLocation) ? union(parPrivateDnsZones, ['privatelink.${varAzBackupGeoCodes[toLower(parLocation)]}.backup.windowsazure.com']) : parPrivateDnsZones
+var varPrivateDnsZonesMerge = contains(varAzBackupGeoCodes, parLocation) ? union(parPrivateDnsZones, [
+  'privatelink.${varAzBackupGeoCodes[toLower(parLocation)]}.backup.windowsazure.com'
+]) : parPrivateDnsZones
```
Comment (jtracey93) [inline infra-as-code/bicep/modules/privateDnsZones/privateDnsZones.bicep:141]: Dont think this should be touched as part of this PR. Think this may be a mistake?

File: infra-as-code/bicep/modules/hubNetworking/hubNetworking.bicep [hunk 3]
```diff
@@ -76,7 +76,15 @@ param parAzFirewallTier string = 'Standard'
...
-param parAzFirewallAvailabilityZones array = []
+param parAzFirewallAvailabilityZones array 
+
+@allowed([
+  '1'
+  '2'
+  '3'
+])
+@description('Availability Zones to deploy the VPN/ER PIP across. Region must support Availability Zones to use. If it does not then leave empty. Ensure that you select a zonal SKU for the ER/VPN Gateway if using Availability Zones for the PIP')
+param parAzGatewayAvailabilityZones array
```
Comment (jtracey93) [inline infra-as-code/bicep/modules/hubNetworking/hubNetworking.bicep:87]: Instead of introducing this as a parameter that is applied to both VPN and ER GWs regardless.

Could we instead add this as a new key to each of the following parameters with the name of `zones` (as an example):
- `parExpressRouteGatewayConfig`
- `parVpnGatewayConfig`

Then this gives the flexibility to configure say the VPN GW without zones and the ER GW with zones (if a customer so chooses).

File: infra-as-code/bicep/modules/hubNetworking/hubNetworking.bicep [hunk 4]
```diff
@@ -179,7 +187,7 @@ param parVpnGatewayConfig object = {
 param parExpressRouteGatewayConfig object = {
   name: '${parCompanyPrefix}-ExpressRoute-Gateway'
   gatewayType: 'ExpressRoute'
-  sku: 'ErGw1AZ'
+  sku: 'ErGw2AZ'
```
Comment (jtracey93) [inline infra-as-code/bicep/modules/hubNetworking/hubNetworking.bicep:190]: Why the change?

File: infra-as-code/bicep/modules/hubNetworking/samples/minimum.sample.bicep [hunk 5]
```diff
@@ -22,5 +22,15 @@ module minimum_hub_network '../hubNetworking.bicep' = {
...
+    parAzFirewallAvailabilityZones: [
+      '1'
+      '2'
+      '3'
+    ]
+    parAzGatewayAvailabilityZones: [
+      '1'
+      '2'
+      '3'
+    ]
```
Comment (jtracey93) [inline infra-as-code/bicep/modules/hubNetworking/samples/minimum.sample.bicep:34]: the minimum sample should not use AZs

**Review by jtracey93 — CHANGES_REQUESTED** [review]

File: infra-as-code/bicep/modules/hubNetworking/samples/minimum.sample.bicep (hunk 5 above)
Comment (jhajduk-microsoft) [reply infra-as-code/bicep/modules/hubNetworking/samples/minimum.sample.bicep]: corrected

File: infra-as-code/bicep/modules/hubNetworking/hubNetworking.bicep (hunk 4 above)
Comment (jhajduk-microsoft) [reply infra-as-code/bicep/modules/hubNetworking/hubNetworking.bicep]: I believe this was for further testing...reverted

File: infra-as-code/bicep/modules/hubNetworking/hubNetworking.bicep (hunk 3 above)
Comment (jhajduk-microsoft) [reply infra-as-code/bicep/modules/hubNetworking/hubNetworking.bicep]: Done. Needs testing

**Comment by jhajduk-microsoft** [comment]
Ready for review tomorrow @jtracey93

File: infra-as-code/bicep/modules/hubNetworking/hubNetworking.bicep (hunk 3 above)
Comment (jhajduk-microsoft) [reply infra-as-code/bicep/modules/hubNetworking/hubNetworking.bicep]: tested. will discuss tomorrow

**Review by jtracey93 — APPROVED** [review]
LGTM

**Review by jtracey93 — APPROVED** [review]

**Comment by jtracey93** [comment]
Merging no breaking changes
